### PR TITLE
Odroid_C2: Switch to linux-amlogic

### PR DESCRIPTION
**Pull request by Kwiboo**
This will make Odroid C2 use LE's linux-amlogic and use similar patches as WeTek Hub.

Kernel config changes need review, my goal was to limit difference between C2 and WH.

This is still in early stage but needs to be tested by more people.

### CODE REVIEW

File: projects/Odroid_C2/linux/linux.aarch64.conf [hunk 1]
```diff
@@ -1077,22 +1086,44 @@ CONFIG_AUDIO_DATA=y
...
+# Amlogic Crypto Support
+#
+CONFIG_CRYPTO_AML=y
+CONFIG_AMLOGIC_WATCHPOINT=y
 
 #
 # Generic Driver Options
 #
-CONFIG_UEVENT_HELPER_PATH=""
+CONFIG_UEVENT_HELPER_PATH="/sbin/mdev"
```
Comment (stefansaraev) [inline projects/Odroid_C2/linux/linux.aarch64.conf]: this one is fine as "", oe does not use uevent helper.

File: projects/Odroid_C2/linux/linux.aarch64.conf [hunk 2]
```diff
@@ -1077,22 +1086,44 @@ CONFIG_AUDIO_DATA=y
...
 CONFIG_DEVTMPFS=y
 CONFIG_DEVTMPFS_MOUNT=y
 CONFIG_STANDALONE=y
 CONFIG_PREVENT_FIRMWARE_BUILD=y
 CONFIG_FW_LOADER=y
 CONFIG_FIRMWARE_IN_KERNEL=y
 CONFIG_EXTRA_FIRMWARE=""
-# CONFIG_FW_LOADER_USER_HELPER is not set
+CONFIG_FW_LOADER_USER_HELPER=y
```
Comment (stefansaraev) [inline projects/Odroid_C2/linux/linux.aarch64.conf]: this is not used and should be disabled.

File: projects/Odroid_C2/linux/linux.aarch64.conf [hunk 3]
```diff
@@ -1207,14 +1231,19 @@ CONFIG_MTD_NAND=y
 # LPDDR flash memory drivers
 #
 # CONFIG_MTD_LPDDR is not set
-# CONFIG_MTD_UBI is not set
+CONFIG_MTD_UBI=y
```
Comment (stefansaraev) [inline projects/Odroid_C2/linux/linux.aarch64.conf]: not sure if this has use case on odroids, but does not harm.

File: projects/Odroid_C2/linux/linux.aarch64.conf [hunk 4]
```diff
@@ -3361,15 +3402,16 @@ CONFIG_HFSPLUS_FS=y
...
 # CONFIG_EFS_FS is not set
 # CONFIG_JFFS2_FS is not set
+# CONFIG_UBIFS_FS is not set
 # CONFIG_LOGFS is not set
 # CONFIG_CRAMFS is not set
 CONFIG_SQUASHFS=y
-CONFIG_SQUASHFS_FILE_CACHE=y
-# CONFIG_SQUASHFS_FILE_DIRECT is not set
-CONFIG_SQUASHFS_DECOMP_SINGLE=y
+# CONFIG_SQUASHFS_FILE_CACHE is not set
```
Comment (stefansaraev) [inline projects/Odroid_C2/linux/linux.aarch64.conf]: keeping squashfs as it was is better, imo. but that's personal opinion

**Comment by lrusak** [comment]
Can the fixup kernel patches be applied to the linux-amlogic repo instead?

File: projects/WeTek_Hub/linux/linux.aarch64.conf [hunk 5]
```diff
@@ -123,18 +125,19 @@ CONFIG_RD_GZIP=y
 # CONFIG_RD_BZIP2 is not set
 # CONFIG_RD_LZMA is not set
 # CONFIG_RD_XZ is not set
-# CONFIG_RD_LZO is not set
+CONFIG_RD_LZO=y
```
Comment (stefansaraev) [inline projects/WeTek_Hub/linux/linux.aarch64.conf]: not needed

File: projects/WeTek_Hub/linux/linux.aarch64.conf [hunk 6]
```diff
@@ -147,16 +150,17 @@ CONFIG_SIGNALFD=y
...
 CONFIG_EMBEDDED=y
 CONFIG_HAVE_PERF_EVENTS=y
 CONFIG_PERF_USE_VMALLOC=y
 
 #
 # Kernel Performance Events And Counters
 #
-# CONFIG_PERF_EVENTS is not set
-# CONFIG_VM_EVENT_COUNTERS is not set
+CONFIG_PERF_EVENTS=y
```
Comment (stefansaraev) [inline projects/WeTek_Hub/linux/linux.aarch64.conf]: any reason for this?

File: projects/WeTek_Hub/linux/linux.aarch64.conf [hunk 7]
```diff
@@ -218,27 +223,9 @@ CONFIG_BLK_DEV_BSGLIB=y
...
-# CONFIG_AMIGA_PARTITION is not set
-# CONFIG_ATARI_PARTITION is not set
-CONFIG_MAC_PARTITION=y
+# CONFIG_PARTITION_ADVANCED is not set
 CONFIG_MSDOS_PARTITION=y
-# CONFIG_BSD_DISKLABEL is not set
-# CONFIG_MINIX_SUBPARTITION is not set
-# CONFIG_SOLARIS_X86_PARTITION is not set
-# CONFIG_UNIXWARE_DISKLABEL is not set
-CONFIG_LDM_PARTITION=y
```
Comment (stefansaraev) [inline projects/WeTek_Hub/linux/linux.aarch64.conf]: windows users may want to attach their ntfs drives via usb enclosures. ldm should not be removed.

File: projects/WeTek_Hub/linux/linux.aarch64.conf [hunk 8]
```diff
@@ -350,7 +339,10 @@ CONFIG_FORCE_MAX_ZONEORDER=11
 #
 # Boot options
 #
-CONFIG_CMDLINE=""
+CONFIG_CMDLINE=" "
```
Comment (stefansaraev) [inline projects/WeTek_Hub/linux/linux.aarch64.conf]: is this tested ?

File: projects/WeTek_Hub/linux/linux.aarch64.conf [hunk 9]
```diff
@@ -934,13 +926,13 @@ CONFIG_AM_VDEC_MPEG12=y
...
 CONFIG_AM_VDEC_H264=y
-# CONFIG_AM_VDEC_H264MVC is not set
+CONFIG_AM_VDEC_H264MVC=y
 CONFIG_AM_VDEC_H264_4K2K=y
 CONFIG_AM_VDEC_H265=y
 CONFIG_AM_VDEC_VP9=y
 CONFIG_AM_VDEC_MJPEG=y
 CONFIG_AM_ENCODER=y
-# CONFIG_AM_JPEG_ENCODER is not set
+CONFIG_AM_JPEG_ENCODER=y
```
Comment (stefansaraev) [inline projects/WeTek_Hub/linux/linux.aarch64.conf]: this probably does nothing anyway.

File: projects/WeTek_Hub/linux/linux.aarch64.conf [hunk 10]
```diff
@@ -1040,10 +1032,11 @@ CONFIG_TVIN_HDMI=y
 # CONFIG_AM_VECM is not set
 # CONFIG_AML_NAND is not set
 CONFIG_AM_INPUT=y
-# CONFIG_AM_SARADC is not set
+CONFIG_AM_SARADC=y
```
Comment (stefansaraev) [inline projects/WeTek_Hub/linux/linux.aarch64.conf]: I dont think we have sar adc on wh board.

**Comment by stefansaraev** [comment]
meh. aaca6843475ae208bca7b19d87c6115af3deed04 changes wetek hub config. cc @codesnake

**Comment by Kwiboo** [comment]
@lrusak kernel patches have been sent to linux-amlogic

The kernel config changes causes issues as I have had a few kernel crashes today, some reboots on the C2 have halted with:
- `Unable to handle kernel paging request at virtual address 3000000000068d7`
- `Unable to handle kernel NULL pointer dereference at virtual address 00000000`
- `init[1]: undefined instruction: pc=ffffffc00111abc0`

**Comment by wrxtasy** [comment]
To note also. I'm getting Cross Compile Badness issues when compiling this version of libamcodec for C2:
https://github.com/LibreELEC/LibreELEC.tv/blob/master/packages/multimedia/libamcodec/package.mk#L29

I use a Codesnake supplied package that compiles without issues:
https://github.com/wrxtasy/LibreELEC.tv.7.0/blob/libreelec-7.1/packages/multimedia/libamcodec/package.mk#L29

File: projects/Odroid_C2/linux/linux.aarch64.conf (hunk 4 above)
Comment (kszaq) [inline projects/Odroid_C2/linux/linux.aarch64.conf]: @stefansaraev Why do you prefer FILE_CACHE over FILE_DIRECT? I'm curious if you've seen performance benefits from the former?

**Comment by chewitt** [comment]
@Kwiboo linux-amlogic changes are merged so please update hashes and rebase, thx

**Comment by Kwiboo** [comment]
This is now rebased and limits the number of kernel config changes.
Patches shared between WH and C2 have been moved to `$PKG_DIR/patches/aarch64/`.

**Comment by wrxtasy** [comment]
I'm of the view that we should seriously think about re-doing the whole boot.ini file for the C2 and be done with it instead of constantly over patching in the update.sh script.  Its not like we have a final release here yet anyways.

I've actually editted out a some junk into a nicer layout, and even that should be trimmed further I believe:
https://github.com/wrxtasy/LibreELEC.tv.7.0/blob/libreelec-7.1/projects/Odroid_C2/bootloader/boot.ini

Also for SoC heat and stability purposes we should stick with a hard and fast CPU Clock speed of 1.536GHz and not give LE C2 users the chance to mess with it, configured as such:

> setenv odroidp2 "${disableuhs} consoleblank=0 max_freq=1536 maxcpus=4"

**Comment by lrusak** [comment]
@wrxtasy I think the best thing to do with the boot.ini is to seperate stuff out for defaults and user defined options, if we can have a default boot.ini that we ship so that we can always make changes to it and a boot.ini that the user can add or change options. This way we can always have a good default for booting and that if the user messes something up then they just need to delete their custom boot.ini file and it will always boot.

**Comment by lrusak** [comment]
I built this PR with jenkins and just tested and it seems to be working great to me :+1:

**Pull request merged**